src/Tokenomia/Vesting/PrivateSale.hs

```diff
@@ -0,0 +1,327 @@
+
+instance Eq Text where 
+  x == y = True
+
+instance Eq B.TxHash where
+  x == y = True
+
+data PrivateSale
+  = PrivateSale
+    { _psAddress         :: B.Address    -- Treasury address
+    , _psStart           :: POSIXTime
+    , _psTranches        :: [Tranche]
+    , _psAssetClass      :: AssetClass   -- Vesting token
+    , _psInvestors       :: [PrivateInvestor]
+    } deriving (Generic, Show)
+
+data Tranche 
+  = Tranche
+    { _tranchePercentage :: Percentage
+    , _trancheDuration   :: Duration 
+    } deriving (Generic, Show)
+
+data PrivateInvestor
+  = PrivateInvestor
+    { _piAddress         :: B.Address
+    , _piAllocation      :: Integer -- Amount of vesting tokens to lock (note, this is likely NOT equal to the total amounts of investments, as different tokens)
+    , _piInvestments     :: [Investment]
+    } deriving (Generic, Show)
+
+data Investment
+  = Investment
+    { _invTx             :: B.TxHash    -- Tx that sends below asset class to payment address
+    , _invAssetClass     :: AssetClass
+    , _invAmount         :: Integer -- Amount of above asset class expected to be sent to payment address
+    } deriving (Generic, Show)
+
+makeLenses ''PrivateSale
+makeLenses ''Tranche
+makeLenses ''PrivateInvestor
+makeLenses ''Investment
+
+verifyPrivateSale :: 
+  (MonadIO m
+  , MonadError  TokenomiaError m
+  , MonadReader Environment m)
+  => m ()
+verifyPrivateSale = do
+  liftIO . putStrLn $ "Please enter a filepath with JSON data"
+  jsonFilePath <- liftIO getLine
+  verifyPrivateSale' jsonFilePath
+
+verifyPrivateSale' :: 
+  (MonadIO m
+  , MonadError  TokenomiaError m
+  , MonadReader Environment m)
+  => FilePath 
+  -> m ()
+verifyPrivateSale' jsonFilePath = do
+  ps <- jsonToPrivateSale jsonFilePath
+  treasAddrTxs <- getTreasAddrTxs ps               -- :: [B.AddressTransaction] <- m [B.AddressTransaction] 
+  let invTxhs = privateSaleToTxhs ps in
+    let txhs = verifyTxHashList invTxhs treasAddrTxs in
+      if length txhs == length invTxhs then do
+        verifyTxs (getPsInvestments ps) txhs
+        else throwError . BlockFrostError . B.BlockfrostError $ "Unequal lengths, means missing TXs"
+
+jsonToPrivateSale ::
+  (MonadIO m
+  , MonadError  TokenomiaError m
+  , MonadReader Environment m)
+  => FilePath 
+  -> m PrivateSale
+jsonToPrivateSale jsonFilePath = do
+    fileContents <- liftIO . readFile $ jsonFilePath
+    case (decode fileContents :: Maybe PrivateSale) of
+        Nothing -> throwError . BlockFrostError . B.BlockfrostError $ "Unable to parse JSON"
+        Just ps -> return ps
+
+privateSaleToTxhs :: PrivateSale -> [B.TxHash]
+privateSaleToTxhs ps = invTxhs 
+
+  where
+    invTxhs :: [B.TxHash]
+    invTxhs = (^. invTx) <$> invs
+
+    invs :: [Investment]
+    invs = getPsInvestments ps
+
+verifyTxs ::
+  (MonadIO m
+  , MonadError  TokenomiaError m
+  , MonadReader Environment m)
+  => [Investment]
+  -> [B.TxHash]
+  -> m ()
+verifyTxs invs = 
+    foldl (\z txh -> do
```

Comment:
You basically _never_ want to use `foldl`. If you really need a left fold (sometimes it's the correct type of fold depending on the circumstance), use `foldl'` (note the prime) from `Data.Foldable`. `foldl` is more or less historical baggage to be avoided